agimus_controller/agimus_controller/trajectory.py

```diff
@@ -37,14 +140,69 @@ class WeightedTrajectoryPoint:
     point: TrajectoryPoint
     weight: TrajectoryPointWeights
 
+    def __eq__(self, other):
+        if not isinstance(other, WeightedTrajectoryPoint):
+            return False
+
+        # Compare the 'point' and 'weight' attributes
+        if self.point != other.point:
+            return False
 
-class TrajectoryBuffer(deque):
+        if self.weight != other.weight:
+            return False
+
+        return True
+
+
+class TrajectoryBuffer(list):
     """List of variable size in which the HPP trajectory nodes will be."""
 
-    def clear_past(self, current_time_ns):
-        while self and self[0].point.time_ns < current_time_ns:
-            self.popleft()
+    def clear_past(self, current_time_ns, dt_ns):
+        while len(self) > 2 and self[0].point.time_ns + 2 * dt_ns < current_time_ns:
+            self.pop(0)
+
+    def find_next_index(self, current_time_ns):
+        """
+        Use bisect_right with a key function to avoid creating a separate list
+        """
+        index = bisect.bisect_right(
+            self, current_time_ns, key=lambda w: w.point.time_ns
+        )
+
+        # Ensure that the index is within bounds
+        if index < len(self):
+            return index
+        else:
+            raise LookupError(
+                "current_time_ns is likely greater than the trajectory horizon time."
+            )
+
+    def compute_horizon_indexes(self, dt_factor_n_seq: list[tuple[int, int]]):
+        indexes = [
+            0,
+        ] * sum(sn for _, sn in dt_factor_n_seq)
+        i = 0
+        for factor, sn in dt_factor_n_seq:
+            for _ in range(sn):
+                if i == 0:
+                    indexes[i] = 0
+                else:
+                    indexes[i] = factor + indexes[i - 1]
+                i += 1
+        # check first time step
+        assert indexes[0] == 0 and "First time step must be 0"
+        # increasing time steps
+        assert all(t0 <= t1 for t0, t1 in zip(indexes[:-1], indexes[1:]))
+        return indexes
 
-    def horizon(self, horizon_size, dt_ocp):
-        # TBD improve this implementation in case the dt_mpc != dt_ocp
-        return self._buffer[: self._ocp.horizon_size]
+    def horizon(self, current_time_ns: int, dt_factor_n_seq: list[tuple[int, int]]):
+        # horizon size
+        horizon_size = sum(sn for _, sn in dt_factor_n_seq)
+        # Starting index:
+        start = self.find_next_index(current_time_ns)
+        indexes = self.compute_horizon_indexes(dt_factor_n_seq)
```

Comment:
I think all those step are too costly to be done at each iterations.
1. the dt_factor_n_seq will not change so one initialization should be sufficient.
2. I think I don't get the rationale of `find_next_index`.